tools/h5part_reader.py

```diff
@@ -0,0 +1,134 @@
+    if type(args.output) == str and args.output != "":
+        if args.output.endswith(".npz"):
+            outfile = args.output
+        else:
+            raise ValueError(f"can only save to NPZ format, not {args.output}")
+
+    with h5py.File(args.input) as file:
+        # Don't use the last timestep,
+        # fluidity deletes the detector and just writes empty data there.
+        n_timesteps = len(file.keys()) - 1
+
+        for particle_id in file["Step#0/id"][:]:
+            option_map = {
+                "olivine": _minerals.MineralPhase.olivine,
+                "enstatite": _minerals.MineralPhase.enstatite,
+                "A": _minerals.OlivineFabric.A,
+                "B": _minerals.OlivineFabric.B,
+                "C": _minerals.OlivineFabric.C,
+                "D": _minerals.OlivineFabric.D,
+                "E": _minerals.OlivineFabric.E,
+                "N": _minerals.EnstatiteFabric.A,
+            }
+
+            # Temporary data arrays.
+            x = np.zeros(n_timesteps)
+            y = np.zeros(n_timesteps)
+            z = np.zeros(n_timesteps)
+            orientations = np.empty((n_timesteps, args.ngrains, 3, 3))
+            fractions = np.empty((n_timesteps, args.ngrains))
+
+            for t in range(n_timesteps):
+
+                # Extract particle position.
+                x[t] = file[f"Step#{t}/x"][particle_id - 1]
+                y[t] = file[f"Step#{t}/y"][particle_id - 1]
+                z[t] = file[f"Step#{t}/z"][particle_id - 1]
+
+                # Extract CPO data.
+                vals = np.empty(args.ngrains * 10)
+                for n in range(len(vals)):
+                    vals[n] = file[f"Step#{t}/CPO{n+1}"][particle_id - 1]
+
+                orientations[t] = np.array(
+                    [
+                        np.reshape(vals[n : n + 9], (3, 3))
+                        for n in range(0, 9 * args.ngrains, 9)
+                    ]
+                )
+                fractions[t] = vals[9 * args.ngrains :]
+
+            _postfix = str(particle_id)
+            _fractions = list(fractions)
+            _orientations = list(orientations)
+            mineral = _minerals.Mineral(
+                phase=option_map[args.fabric],
+                fabric=option_map[args.phase],
+                n_grains=args.ngrains,
+                fractions_init=_fractions[0],
+                orientations_init=_orientations[0],
+            )
+            mineral.fractions = _fractions
+            mineral.orientations = _orientations
+            mineral.save(outfile, postfix=_postfix)
+            archive = ZipFile(outfile, mode="a", allowZip64=True)
+            for key, data in zip(("x", "y", "z"), (x, y, z)):
+                with archive.open(f"{key}_{_postfix}", "w", force_zip64=True) as file:
+                    buffer = io.BytesIO()
+                    np.save(buffer, data)
+                    file.write(buffer.getvalue())
+                    buffer.close()
```

Comment:
Why not use [`numpy.savez`](https://numpy.org/doc/stable/reference/generated/numpy.savez.html)?

Reply:
It doesn't support appending, so that would overwrite the result of `mineral.save()` from two lines above.